src/main/resources/assets/computercraft/lua/rom/programs/http/gist.lua

```diff
@@ -0,0 +1,335 @@
+        if handle.getResponseCode() ~= 200 then
+            handle.close()
+            error("Failed to download file data.")
+        end
+        local d = handle.readAll()
+        handle.close()
+        return d
+    end
+end
+
+local function setTextColor(c) if term.isColor() then term.setTextColor(c) elseif c == colors.white or c == colors.yellow then term.setTextColor(colors.white) else term.setTextColor(colors.lightGray) end end
+
+local function requestAuth(headers, interactive)
+    if settings.get("gist.id") ~= nil then headers.Authorization = "token " .. settings.get("gist.id") elseif interactive then
+        setTextColor(colors.yellow)
+        write("You need to add a Personal Access Token (PAK) to upload Gists. Follow the instructions at ")
+        setTextColor(colors.blue)
+        write("https://tinyurl.com/GitHubPAK")
+        setTextColor(colors.yellow)
+        write(" to generate one. Make sure to check the '")
+        setTextColor(colors.blue)
+        write("gist")
+        setTextColor(colors.yellow)
+        print("' checkbox on step 7 (under 'Select scopes'). Once done, paste it here.")
+        setTextColor(colors.lime)
+        write("PAK: ")
+        setTextColor(colors.white)
+        local pak = read()
+        if pak == nil or pak == "" then error("Invalid PAK, please try again.") end
+        settings.set("gist.id", pak)
+        headers.Authorization = "token " .. pak
+    end
+end
+
+-- User API - this can be loaded with os.loadAPI
+
+-- ID can be either just the gist ID or a gist ID followed by a slash and a file name
+-- * If a file name is specified, retrieves that file
+-- * Otherwise, if there's only one file, retrieves that file
+-- * Otherwise, if there's a file named 'init.lua', retrieves 'init.lua'
+-- * Otherwise, if there's more than one file but only one *.lua file, retrieves the Lua file
+-- * Otherwise, retrieves the first Lua file alphabetically (with a warning)
+-- * Otherwise, fails
+
+--- Retrieves one file from a Gist using the specified ID.
+-- @tparam string id The Gist ID to download from. See above comments for more details.
+-- @tparam[opt] any showMessages Set to any true value to show messages while connecting.
+-- @treturn string|nil The contents of the specified Gist file, or nil on error.
+-- @treturn string|nil The name of the file that was chosen to be downloaded, or nil on error.
+function get(id, showMessages)
+    expect(1, id, "string")
+    local file
+    if id:find("/") ~= nil then id, file = id:match("^([0-9A-Fa-f:]+)/(.+)$") end
+    if id == nil or not id:match("^[0-9A-Fa-f][0-9A-Fa-f:]+[0-9A-Fa-f]$") then error("bad argument #1 to 'get' (invalid ID)", 2) end
+    if id:find(":") ~= nil then id = id:gsub(":", "/") end
+    if showMessages then write("Connecting to api.github.com... ") end
+    local handle = http.get("https://api.github.com/gists/" .. id)
+    if handle == nil then if showMessages then print("Failed.") end return nil end
```

Comment:
Just a couple of things:

 - It's probably nicer to return `nil, <failure reason>` rather than displaying a message here. That's more in line with what's done elsewhere.
 - Can we attempt to extract a reason from the response message and/or handle? For instance, "Cannot connect", or the JSON `message` field.

Reply:
I've fixed the functions so they now return `nil` + an error message. They also print `Failed.` through `progress` if specified. I also made it so the message is added if available.